README.md
# Net-Util
A small python command line tool used to automatically handle the TPC connection when using scapy

## Usage
To start the program please run the following command
```angular2html
sudo python3 net_util.py 
```

## Notice
this tool my fail to connect unless you run the following command
```
    iptables -A OUTPUT -p tcp --tcp-flags RST RST -s <INSERT IP> -j DROP
```
The reason is that the kernel is unaware of what scapy is doing and sends a RST packet when it sees incoming packets

changelog.txt
v0.3 changelog

- Added more print outs for verbose
- Added the ability to save the connection config

net_util.py
import readline
import signal
import sys

from connection import Connection
from scapy.all import *

VERSION = '0.3'

# iptables -A OUTPUT -p tcp --tcp-flags RST RST -s 192.168.1.20 -j DROP

connection = Connection()


def config(**kwargs):
    connection.config(**kwargs)


def load(module):
    load_contrib(module)


def show(pkt):
    pkt.show()


def connect(**kwargs):
    connection.connect(**kwargs)


def disconnect(**kwargs):
    connection.disconnect(**kwargs)


def send(payload, **kwargs):
    connection.send(payload, **kwargs)


def save():
    connection.save()


def reset(**kwargs):
    connection.reset(**kwargs)


def help():
    print('''
    connect() - connect to server
    disconnect() - disconnect from server
    load(<contrib name>) - loads contrib module
    save() - saves the application configuration
    send(<packet>) - sends packet(s)
    show(<scapy packet>) - shows a scapy packet
    reset() - resets the connection
    ''')


def exit():
    if connection.is_connected():
        print("\nsending RST packet to open connection")
        connection.reset()
    print("\nGoodbye")
    sys.exit()


def sigint_handler(sig, frame):
    exit()


if __name__ == '__main__':
    signal.signal(signal.SIGINT, sigint_handler)
    print("net-util v{}".format(VERSION))
    while True:
        val = input(">>>")
        if val == '':
            continue
        try:
            exec(val)
        except Exception as e:
            print(e)


connection.old.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
# Author: N0dr4x (n0dr4x@protonmail.com)

'''
Simple Scapy TCP Session class that provide ability
to : 
   - execute the 3-way handshake (eg. connect)
   - properly close connection (->FIN/ACK, <-FIN/ACK, ->ACK )
   - send automatic acknowledgment of received tcp data packet
   - build a next packet to send with correct sequence number
   - directly send data through the session
HINT : Don't forget to block TCP/RST packet that was send
       by the linux kernel because no source port was bound.

         # iptables -A OUTPUT -p tcp --sport 1337 --tcp-flags RST RST -j DROP
Source port is, for now, fixed to 1337 to facilitate wireshark filtering.
The purpose of this class is to easily build a working tcp session and
have complete scapy control of the next tcp packet.
Usage & example :

   # # Create the session object and connect to host 192.168.13.37 port 80
   # >>> sess = TcpSession(('192.168.13.37',80))
   # >>> sess.connect()
   # # Build next packet and send it fragmented (layer 2)
   # >>> p = sess.build('GET / HTTP/1.1\r\n\r\n')
   # >>> send(fragment(p, fragsize=16))
   # # Direct send data through the session and close
   # >>> sess.send('GET /index.html HTTP/1.1\r\n\r\n')
   # >>> sess.close()
   # # Session object can be reusable
   # >>> sess.connect()
   # >>> sess.send('GET /robot.txt HTTP/1.1\r\n\r\n')
   # >>> sess.close()
TODO :
   1/ Optionally dump received data to a file
   2/ Proper logging
'''

from scapy.all import *
from threading import Thread

from scapy.contrib.bgp import BGP, BGPUpdate
from scapy.layers.inet import TCP, IP


class TcpSession:

    def __init__(self):
        self.seq = 0
        self.ack = 0
        self.ip = IP(dst="192.168.59.129")
        self.sport = 1337
        self.dport = 179
        self.connected = False
        self._ackThread = None
        self._timeout = 10

    def _ack(self, p):
        self.ack = p[TCP].seq + len(p[Raw])
        ack = self.ip / TCP(sport=self.sport, dport=self.dport, flags='A', seq=self.seq, ack=self.ack)
        send(ack)

    def _ack_rclose(self):
        self.connected = False

        self.ack += 1
        fin_ack = self.ip / TCP(sport=self.sport, dport=self.dport, flags='FA', seq=self.seq, ack=self.ack)
        ack = sr1(fin_ack, timeout=self._timeout)
        self.seq += 1

        assert ack.haslayer(TCP), 'TCP layer missing'
        assert ack[TCP].flags & 0x10 == 0x10, 'No ACK flag'
        assert ack[TCP].ack == self.seq, 'Acknowledgment number error'

    def _sniff(self):
        s = L3RawSocket()
        while self.connected:
            p = s.recv(MTU)
            if p.haslayer(TCP) and p.haslayer(Raw) \
                    and p[TCP].dport == self.sport:
                self._ack(p)
            if p.haslayer(TCP) and p[TCP].dport == self.sport \
                    and p[TCP].flags & 0x01 == 0x01:  # FIN
                self._ack_rclose()

        s.close()
        self._ackThread = None
        print('Acknowledgment thread stopped')

    def _start_ackThread(self):
        self._ackThread = Thread(name='AckThread', target=self._sniff)
        self._ackThread.start()

    def connect(self):
        self.seq = random.randrange(0, (2 ** 32) - 1)

        syn = self.ip / TCP(sport=self.sport, dport=self.dport, seq=self.seq, flags='S')
        syn_ack = sr1(syn, timeout=self._timeout)
        self.seq += 1

        assert syn_ack.haslayer(TCP), 'TCP layer missing'
        assert syn_ack[TCP].flags & 0x12 == 0x12, 'No SYN/ACK flags'
        assert syn_ack[TCP].ack == self.seq, 'Acknowledgment number error'

        self.ack = syn_ack[TCP].seq + 1
        ack = self.ip / TCP(sport=self.sport, dport=self.dport, seq=self.seq, flags='A', ack=self.ack)
        send(ack)

        self.connected = True
        self._start_ackThread()
        print('Connected')

    def close(self):
        self.connected = False

        fin = self.ip / TCP(sport=self.sport, dport=self.dport, flags='FA', seq=self.seq, ack=self.ack)
        fin_ack = sr1(fin, timeout=self._timeout)
        self.seq += 1

        assert fin_ack.haslayer(TCP), 'TCP layer missing'
        assert fin_ack[TCP].flags & 0x11 == 0x11, 'No FIN/ACK flags'
        assert fin_ack[TCP].ack == self.seq, 'Acknowledgment number error'

        self.ack = fin_ack[TCP].seq + 1
        ack = self.ip / TCP(sport=self.sport, dport=self.dport, flags='A', seq=self.seq, ack=self.ack)
        send(ack)

        print('Disconnected')

    def build(self, payload):
        psh = self.ip / TCP(sport=self.sport, dport=self.dport, flags='PA', seq=self.seq, ack=self.ack) / payload
        print(psh)
        self.seq += len(raw(psh))
        return psh

    def send(self, payload):
        psh = self.build(payload)
        ack = sr1(psh, timeout=self._timeout)

        assert ack.haslayer(TCP), 'TCP layer missing'
        assert ack[TCP].flags & 0x10 == 0x10, 'No ACK flag'
        assert ack[TCP].ack == self.seq, 'Acknowledgment number error'


if __name__ == '__main__':
    tcp_session = TcpSession()
    tcp_session.connect()
    tcp_session.send(BGPUpdate())
    tcp_session.close()

import random
import time

from scapy.contrib.bgp import BGPUpdate, BGP
from scapy.layers.inet import IP, TCP
from scapy.sendrecv import sr1, send
#
# dport = 179
# sport = random.randint(30000, 65000)
# src = '192.168.59.128'
# dst = '192.168.59.129'
#
# # SYN
# ip = IP(src=src, dst=dst)
# SYN = TCP(sport=sport, dport=dport, flags='S', seq=1000)
# SYNACK = sr1(ip / SYN)
#
# # SYN-ACK
# ACK = TCP(sport=sport, dport=dport, flags='A', seq=SYNACK.ack, ack=SYNACK.seq+1)
# send(ip / ACK)
#
# send(ip / TCP(sport=sport, dport=dport, flags="PA", seq=SYNACK.ack, ack=SYNACK.seq+1) / BGP())
# time.sleep(1)
# send(ip / TCP(sport=sport, dport=dport, flags="R"))
#
#
# FIN = ip / TCP(sport=sport, dport=dport, flags="FA", seq=SYNACK.ack, ack=SYNACK.seq + 1)
# FINACK = sr1(FIN, timeout=1, multi=1)
#
# # assert FINACK[1].haslayer(TCP), 'TCP layer missing'
# # assert FINACK[1][TCP].flags & 0x11 == 0x11, 'No FIN/ACK flags'
#
# LASTACK = ip / TCP(sport=sport, dport=dport, flags="A", seq=FINACK[0].ack, ack=FINACK[0].seq)
# send(LASTACK)


connection.py
import configparser
import threading

from scapy.all import *
from scapy.layers.inet import IP, TCP

CONFIG_FILE = 'config.ini'


# TODO: finish verbose outputs
# TODO: set timeout in send calls
# TODO: add log saving
class Connection:

    def __init__(self):
        if not os.path.isdir('logs'):
            os.mkdir('logs')

        config = configparser.ConfigParser()
        config.read(CONFIG_FILE)

        # application variables
        self.src = str(config['APP_CONFIG']['src']) if config.has_option('APP_CONFIG', 'src') else '1.1.1.1'
        self.dst = str(config['APP_CONFIG']['dst']) if config.has_option('APP_CONFIG', 'dst') else '1.1.1.1'
        self.dport = int(config['APP_CONFIG']['dport']) if config.has_option('APP_CONFIG', 'dport') else 10000
        self.sport = int(config['APP_CONFIG']['sport']) if config.has_option('APP_CONFIG', 'sport') else 10000
        self.timeout = int(config['APP_CONFIG']['timeout']) if config.has_option('APP_CONFIG', 'timeout') else 5
        self.base_seq = int(config['APP_CONFIG']['base_seq']) if config.has_option('APP_CONFIG', 'base_seq') else 1000
        v_string = config['APP_CONFIG']['verbose'] if config.has_option('APP_CONFIG', 'verbose') else 'False'
        self.v = True if v_string == 'True' else False

        # session/connection variables
        self.ip = None
        self.connected = False
        self.seq = self.base_seq
        self.base_ack = 0
        self.ack = self.base_ack

        #  multithreading
        self._receiving_thread = None
        self._lock = threading.Lock()
        self._padding = True

    def config(self, src=None, dst=None, sport=None, dport=None,
               timeout=None, base_seq=None, seq=None, ack=None, v=None):
        self.src = src if src else self.src
        self.dst = dst if dst else self.dst
        self.sport = sport if sport else self.sport
        self.dport = dport if dport else self.dport
        self.timeout = timeout if timeout else self.timeout
        self.base_seq = base_seq if base_seq else self.base_seq
        self.seq = seq if seq else self.seq
        self.ack = ack if ack else self.ack
        self.v = v if v is not None else self.v

        print('''
    ==== APPLICATION CONFIGURATION ====
    source ip(src):             {}
    destination ip(dst):        {}
    source port(sport):         {}
    destination port(dport):    {}
    timeout(timeout):           {}
    base seq number(base_seq):  {}
    verbose(v):                 {}
    
    ==== SESSION DATA ====
    connection status           {}
    current seq number(seq)     {}({})
    base ack number:            {}
    current ack number(ack):    {}({})
        '''.format(
            self.src,
            self.dst,
            self.sport,
            self.dport,
            self.timeout,
            self.base_seq,
            self.v,
            True if self.connected else False,
            self.seq,
            self.seq - self.base_seq,
            self.base_ack,
            self.ack,
            self.ack - self.base_ack
        ))

    def is_connected(self):
        return self.connected

    def connect(self, v=None):
        verbose = self.v if v is None else v

        if self.connected:
            print('ERROR YOU ARE CURRENTLY CONNECTED')
            return

        try:
            # SYN
            self.ip = IP(src=self.src, dst=self.dst)
            syn = self.ip / TCP(sport=self.sport, dport=self.dport, flags='S', seq=self.seq)

            if verbose:
                print("============= SYN PACKET =============")
                syn.show()
                print("=======================================")

            syn_ack = sr1(syn, timeout=self.timeout)

            if verbose:
                print("============== RESPONSE ==============")
                syn_ack.show()
                print("=======================================")

            assert syn_ack.haslayer(TCP), 'TCP layer missing'
            assert syn_ack[TCP].flags & 0x12 == 0x12, 'No SYN/ACK flags'

            # Updating seq and ack numbers
            self.seq += 1
            self.base_ack = syn_ack.seq
            self.ack = self.base_ack
            self.ack += 1

            assert syn_ack[TCP].ack == self.seq, 'Acknowledgment number error'

            # sending ack response
            ack = self.ip / TCP(sport=self.sport, dport=self.dport, flags='A', seq=self.seq, ack=self.ack)

            if verbose:
                print("============ ACK PACKET ============")
                syn_ack.show()
                print("=======================================")

            send(ack)

            self.connected = True
            self._receiving_thread = threading.Thread(target=self._receiving_thread_func, args=())
            self._receiving_thread.start()

        except Exception as ex:
            print(ex)
            print("FAILED TO CONNECT, SENDING RESET")
            self.reset()

    # TODO: fix disconnect
    def disconnect(self, v=None):
        verbose = self.v if v is None else v

        self._lock.acquire()
        try:
            fin = self.ip / TCP(sport=self.sport, dport=self.dport, flags="FA", seq=self.seq, ack=self.ack)
            ack_fin_ack = sr1(fin, timeout=self.timeout, multi=1)
            ack = ack_fin_ack[0]
            fin_ack = ack_fin_ack[1]

            assert fin_ack.haslayer(TCP), 'TCP layer missing'
            assert fin_ack[TCP].flags & 0x11 == 0x11, 'No FIN/ACK flags, received \'{}\''.format(fin_ack[TCP].flags)

            last_ack = self.ip / TCP(sport=self.sport, dport=self.dport, flags="A", seq=ack.seq,
                                     ack=ack.ack)
            send(last_ack)
            self._receiving_thread.join()
            self.connected = False
            self._receiving_thread.join()
        except Exception as ex:
            print(ex)
            print("FAILED TO DISCONNECT, SENDING RESET")
            self.reset()
        finally:
            self._lock.release()

    def log(self, inbound):
        pass

    def reset(self, seq=None, v=None):
        verbose = self.v if v is None else v

        self._lock.acquire()
        try:
            seq = seq if seq else self.seq
            ip = IP(src=self.src, dst=self.dst)
            rst = ip / TCP(sport=self.sport, dport=self.dport, flags="R", seq=seq)

            if verbose:
                print("=========== RESET PACKET ===========")
                rst.show()
                print('====================================')
            send(rst)

            self.base_ack = 0
            self.ack = 0
            self.seq = self.base_seq
            self.connected = False
            if self._receiving_thread:
                self._receiving_thread.join()

        except Exception as ex:
            print(ex)
            print("FAILED TO SEND RESET")
        finally:
            self._lock.release()

    def save(self):
        config = configparser.ConfigParser()
        config['APP_CONFIG'] = {
            'src': self.src,
            'dst': self.dst,
            'sport': self.sport,
            'dport': self.dport,
            'timeout': self.timeout,
            'base_seq': self.seq,
            'verbose': self.v
        }
        with open(CONFIG_FILE, 'w') as config_file:
            config.write(config_file)
            print("Configuration saved")

    def send(self, payload, v=None):
        verbose = self.v if v is None else v

        if self.connected is False:
            print("ERROR, not connected")
            return

        self._lock.acquire()
        try:
            pkt = self.ip / TCP(sport=self.sport, dport=self.dport, flags="PA", seq=self.seq, ack=self.ack) / payload

            if verbose:
                print("=========== SENDING PACKET ===========")
                pkt.show()
                print("=======================================")

            ack = sr1(pkt, timeout=self.timeout, verbose=False)
            self.seq += len(payload)

            if verbose:
                print("============== RESPONSE ==============")
                ack.show()
                print("=======================================")

            assert ack.haslayer(TCP), 'TCP layer missing'
            assert ack[TCP].flags & 0x10 == 0x10, 'No ACK flag'

        except Exception as ex:
            print(ex)
            print("FAILED TO SEND PAYLOAD")

        finally:
            self._lock.release()

    def _receiving_thread_func(self):
        while self.connected:
            sniff(filter=' tcp and src host {} and port {}'.format(self.dst, self.sport), count=2,
                  prn=self._ack, timeout=1)

    def _ack(self, pkt):
        if self._padding:
            self._padding = False
            return
        else:
            self._padding = True

        print("============== RECEIVED ==============")
        pkt.show()
        print("=======================================")

        self._lock.acquire()
        try:
            self.ack += len(pkt[TCP].load)
            ack = self.ip / TCP(sport=self.sport, dport=self.dport, flags='A', seq=self.seq, ack=self.ack)
            send(ack, verbose=False)
        except Exception as ex:
            print(ex)
        finally:
            self._lock.release()

